src/bits32/segmentation.rs

```diff
@@ -0,0 +1,40 @@
+use core::mem::size_of;
+
+use bits32::task::*;
+use shared::descriptor;
+use shared::PrivilegeLevel;
+pub use shared::segmentation::*;
+
+/// Reload code segment register.
+/// Note this is special since we can not directly move
+/// to %cs. Instead we push the new segment selector
+/// and return value on the stack and use lretl
+/// to reload cs and continue at 1:.
+pub unsafe fn set_cs(sel: SegmentSelector) {
+    asm!("pushl $0; \
+          pushl $$1f; \
+          lretl; \
+          1:" :: "ri" (sel.bits() as usize) : "memory");
+}
+
+impl SegmentDescriptor {
+    pub fn new_memory(base: u32, limit: u32, ty: Type, accessed: bool, dpl: PrivilegeLevel) -> SegmentDescriptor {
+        let ty1 = descriptor::Type::SegmentDescriptor {
+            ty: ty,
+            accessed: accessed,
+        };
+        let flags = FLAGS_DB;
+        let seg = SegmentDescriptor::memory_or_tss(base, limit, ty1, dpl, flags);
+        seg
+    }
+
+    pub fn new_tss(tss: &TaskStateSegment, dpl: PrivilegeLevel) -> [SegmentDescriptor; 2] {
+        let tss_ptr = tss as *const TaskStateSegment;
+        let ty1 = descriptor::Type::SystemDescriptor {
+            size: true,
+            ty: descriptor::SystemType::TssAvailable,
+        };
+        let seg = SegmentDescriptor::memory_or_tss(tss_ptr as u32, size_of::<TaskStateSegment>() as u32, ty1, dpl, Flags::empty());
+        seg
```

Comment:
The function signature declares that it will return an array of two `SegmentDescriptor`s but instead it returns one segment.

```bash
error[E0308]: mismatched types
  --> /home/rob/.cargo/git/checkouts/rust-x86-3fe08cbd15a995a6/18aee0a/src/bits32/segmentation.rs:38:9
   |
38 |         seg
   |         ^^^ expected array of 2 elements, found struct `shared::segmentation::SegmentDescriptor`
   |
   = note: expected type `[shared::segmentation::SegmentDescriptor; 2]`
              found type `shared::segmentation::SegmentDescriptor`
```

Reply:
Thanks for noticing! Fixed the 32-bit version in my latest commit.

Reply:
For your information, further work on rust-x86 is now continued at https://github.com/RWTH-OS/rust-x86/tree/all-merged
My fork is only kept alive until the Pull Requests have been merged.